instrumentation/opentelemetry-instrumentation-openai-v2/src/opentelemetry/instrumentation/openai_v2/utils.py

```diff
@@ -43,69 +48,138 @@ def wrapper(*args, **kwargs):
+def extract_content_from_prompt_message(message) -> Tuple[str, Any]:
+    content = get_property_value(message, "content")
+    if content:
+        return ("content", content)
+
+    tool_calls = extract_tool_calls(message)
+    if tool_calls:
+        return ("tool_calls", tool_calls)
 
-    # Return an empty string if none of the above conditions are met
     else:
-        return ""
+        return None
 
 
-def extract_tools_prompt(item):
-    tool_calls = getattr(item, "tool_calls", None)
+def extract_tool_calls(item):
+    tool_calls = get_property_value(item, "tool_calls")
     if tool_calls is None:
         return
 
     calls = []
     for tool_call in tool_calls:
         tool_call_dict = {
-            "id": getattr(tool_call, "id", ""),
-            "type": getattr(tool_call, "type", ""),
+            "id": get_property_value(tool_call, "id"),
+            "type": get_property_value(tool_call, "type"),
         }
 
-        if hasattr(tool_call, "function"):
-            tool_call_dict["function"] = {
-                "name": getattr(tool_call.function, "name", ""),
-                "arguments": getattr(tool_call.function, "arguments", ""),
-            }
+        func = get_property_value(tool_call, "function")
+        if func:
+            tool_call_dict["function"] = {}
+
+            name = get_property_value(func, "name")
+            if name:
+                tool_call_dict["function"]["name"] = name
+
+            arguments = get_property_value(func, "arguments")
+            if arguments:
+                tool_call_dict["function"]["arguments"] = arguments.replace(
+                    "\n", ""
+                )
+
         calls.append(tool_call_dict)
     return calls
 
 
-def set_event_prompt(span, prompt):
-    span.add_event(
-        name="gen_ai.content.prompt",
-        attributes={
-            GenAIAttributes.GEN_AI_PROMPT: prompt,
-        },
+def set_server_address_and_port(client_instance, attributes):
+    base_client = getattr(client_instance, "_client", None)
+    base_url = getattr(base_client, "base_url", None)
+    if not base_url:
+        return
+
+    port = -1
+    if isinstance(base_url, URL):
+        attributes[ServerAttributes.SERVER_ADDRESS] = base_url.host
+        port = base_url.port
+    elif isinstance(base_url, str):
+        try:
+            url = urlparse(base_url)
+            attributes[ServerAttributes.SERVER_ADDRESS] = url.hostname
+            port = url.port
+        except Exception:
+            pass
+
+    if port and port != 443 and port > 0:
+        attributes[ServerAttributes.SERVER_PORT] = port
+
+
+def get_property_value(obj, property_name):
+    if isinstance(obj, dict):
+        return obj.get(property_name, None)
+
+    return getattr(obj, property_name, None)
+
+
+def message_to_event(message, span_ctx):
+    attributes = {
+        GenAIAttributes.GEN_AI_SYSTEM: GenAIAttributes.GenAiSystemValues.OPENAI.value
+    }
+    role = get_property_value(message, "role")
```

Comment:
Just curious, are there `role` values that are not currently defined in sem conv? (system, user, assistant, tool, etc)?

Reply:
not that I'm aware of, but I believe users can pass any role and it may work in some edge cases or with some non-OpenAI backends (a lot of backends are Open-AI compatible on the REST level and this instrumentation can be used with any of them).